### PR TITLE
Feat/trading form validation

### PR DESCRIPTION
**Pull request by da-kami**
resolves https://github.com/get10101/10101/issues/131

Please read this:

We keep seeing an exception thrown by flutter in the logs.

TLDR: 
I would just ignore the exception thrown by Flutter. Nothing actually breaks in the UI, but Flutter keeps complaining. The user should never see this, and I don't think it's a big deal...

---

Long story:

After making the `TradeBottomSheetTab` stateful we keep seeing this exception in the logs (when modifying the `margin`; we also see it when modifying the `quantity`, but slightly different because it is always triggered by the other text field):

```
══╡ EXCEPTION CAUGHT BY FOUNDATION LIBRARY ╞════════════════════════════════════════════════════════
The following assertion was thrown while dispatching notifications for TextEditingController:
setState() or markNeedsBuild() called during build.
This Form widget cannot be marked as needing to build because the framework is already in the
process of building widgets. A widget can be marked as needing to be built during the build phase
only if one of its ancestors is currently building. This exception is allowed because the framework
builds parent widgets before children, which means a dirty descendant will always be built.
Otherwise, the framework might not visit this widget during this build phase.
The widget on which setState() or markNeedsBuild() was called was:
  Form-[LabeledGlobalKey<FormState>#3fa9e]
The widget which was currently being built when the offending call was made was:
  DoubleTextInputFormField

When the exception was thrown, this was the stack:

...
```
When using `TextEdigintController`s and a `Form` it is recommended to have a `Stateful` widget.

Part of the problem comes from the complexity of having two text fields that update each other - sounds easy, but is actually quite complex. But even without editing both fields we can run into the exception of triggering multiple builds. 

I added `print` statements to see how often certain code pathes (onChange, and build) are triggered and they are triggered too often - but I just don't understand how. In many cases I did not even interact with the UI and somehow it triggered a rebuild. I am not sure what in our setup triggers this, but something is and I don't have the time to look into this now.

### CODE REVIEW

**Comment by klochowicz** [comment]
it sounds like there's an architectural issue, ie. we're not using Flutter correctly 😞 Fixing it might be a a rabbit hole that takes quite a bit of time....

Whilst I don't think it's a blocker launching the beta, but it would be best to prioritise fixing this issue shortly afterwards. 
it might have side effects like:
-  flooding the logs (space on the device) and harder time for our debugging
- app stability (maybe the UI responsiveness is affected?)

File: mobile/lib/features/trade/trade_bottom_sheet_tab.dart [hunk 1]
```diff
@@ -79,7 +80,19 @@ class _TradeBottomSheetTabState extends State<TradeBottomSheetTab> {
...
                     const Flexible(child: Text("Usable Balance:")),
                     const SizedBox(width: 5),
-                    Flexible(child: AmountText(amount: Amount(usableBalance)))
+                    Flexible(child: AmountText(amount: Amount(usableBalance))),
+                    const SizedBox(
+                      width: 5,
+                    ),
+                    ModalBottomSheetInfo(
+                      infoText:
+                          "Your usable balance of $usableBalance sats takes a fixed reserve of ${provider.reserve} sats into account. "
```
Comment (klochowicz) [inline mobile/lib/features/trade/trade_bottom_sheet_tab.dart:91]: curious - why is `$usableBalance` without `{}` when others are?

**Review by klochowicz — APPROVED** [review]
LGTM! 

does it still trigger the exceptions with the additional commits you've added?

**Comment by da-kami** [comment]
> LGTM!
> 
> does it still trigger the exceptions with the additional commits you've added?

Unfortunately it does - the problem must be related to using the providor's `Selector` in combination with the `FormTextFields` and validation. I stopped investigating because it's too time intense. 
I created https://github.com/get10101/10101/issues/402 that sums up the problem and adds my recommendation :)

File: mobile/lib/features/trade/application/trade_values_service.dart [hunk 2]
```diff
@@ -24,4 +24,21 @@ class TradeValuesService {
...
+  int getChannelReserve() {
+    // TODO: Fetch from backend
+    // This is the minimum value that has to remain in the channel. It is defined in rust-lightning and we should fetch this value from the corresponding constant in the backend.
+    return 1000;
+  }
+
+  int getMinTradeMargin() {
+    // This value is an arbitrary number; we only allow trades with a minimum of 1000 sats margin.
+    return 1000;
+  }
```
[the remark this answers is not in the document]
Comment (da-kami) [reply mobile/lib/features/trade/application/trade_values_service.dart:43]: Some of these values are actually not dummy values, but they should be derived based on other values in the backend. I just did not pull that through to move faster 😅

I will go on implementing things, but I'll add a ticket to pull it through - want to avoid spending more time on this.

[the remark this answers is not in the document]
Comment (klochowicz) [reply mobile/lib/features/trade/application/trade_values_service.dart:43]: I think that's the definition of all dummy values - we've put them there just to move faster for now 😂
